File: webdriverio-ui-demo/src/helpers/ElementHelper.ts
import type { ChainablePromiseElement } from 'webdriverio';

/**
 * Cross-site helper for operations WebdriverIO doesn't expose natively.
 *
 * Rule of thumb: try the native WebdriverIO API first
 *   - element.click()                          → before jsClick
 *   - element.scrollIntoView()                 → before jsScrollIntoView
 *   - element.getCSSProperty(prop)              → covers getComputedStyle natively (kept
 *                                                  here only as a thin wrapper for call-site
 *                                                  consistency with the other helpers)
 *   - (no native "open target=_blank in place")  → openSameTab
 *
 * Only fall back to jsClick/jsScrollIntoView when the native API genuinely
 * doesn't fit (e.g. overlay intercepts the click, link has target="_blank"
 * and you want the navigation in-place).
 *
 * The helper takes no constructor arguments — every method operates on a
 * `ChainablePromiseElement`, WebdriverIO's lazy element handle.
 */
export class ElementHelper {
  /**
   * Click via JS `element.click()`, bypassing WebdriverIO's actionability
   * checks. Useful when an animation or overlay intercepts a real click but
   * the element is functionally present.
   *
   * Why non-native: WebdriverIO's `click()` scrolls into view and waits for
   * the element to be clickable; there is no built-in escape hatch short of
   * dropping to a raw DOM click.
   */
  async jsClick(el: ChainablePromiseElement): Promise<void> {
    await browser.execute((elem) => (elem as HTMLElement).click(), await el);
  }

  /**
   * Scroll the element to the centre of the viewport.
   *
   * Why non-native: WebdriverIO's `scrollIntoView()` defaults to
   * `block: 'start'`, which often leaves elements pinned under a sticky
   * header or footer.
   */
  async jsScrollIntoView(
    el: ChainablePromiseElement,
    block: ScrollLogicalPosition = 'center',
  ): Promise<void> {
    await browser.execute(
      (elem, b) => elem.scrollIntoView({ block: b as ScrollLogicalPosition, inline: 'center' }),
      await el,
      block,
    );
  }

  /** Read a single resolved CSS property from an element. Native since WDIO's
   * `getCSSProperty` already parses the computed value — this wrapper exists
   * only so every helper is called the same way from page objects. */
  async getComputedStyle(el: ChainablePromiseElement, property: string): Promise<string> {
    const result = await el.getCSSProperty(property);
    return String(result.value);
  }

  /**
   * Click a link that has `target="_blank"` and have the navigation happen
   * in the current tab rather than spawning a new window.
   *
   * Why non-native: WebdriverIO's native pattern for a new tab is
   * `browser.getWindowHandles()` / `browser.switchToWindow()`, which forces
   * every test that follows such a link to juggle window handles. For sites
   * where this is pervasive (e.g. WebDriverUniversity's challenge cards),
   * stripping `target` keeps tests linear.
   */
  async openSameTab(link: ChainablePromiseElement): Promise<void> {
    await browser.execute((el) => el.removeAttribute('target'), await link);
    await link.click();
  }
}

File: selenium-java-ui-demo/src/test/java/com/demo/helpers/ElementHelper.java
package com.demo.helpers;

import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

/**
 * A small set of escape hatches for things Selenium's native API doesn't cover, or doesn't cover
 * well enough for every page object to reimplement. Mirrors the sibling Playwright/WebdriverIO
 * repos' ElementHelper: four methods, added only because the native option falls short AND
 * multiple page objects need it. Always try the native option first — see the table in
 * docs/CODEBASE_GUIDE.md.
 */
public class ElementHelper {

  private final WebDriver driver;

  public ElementHelper(WebDriver driver) {
    this.driver = driver;
  }

  /** Bypasses Selenium's actionability checks when an overlay/animation intercepts a real click. */
  public void jsClick(WebElement element) {
    ((JavascriptExecutor) driver).executeScript("arguments[0].click();", element);
  }

  /**
   * Selenium has no native scrollIntoView on WebElement (unlike Playwright/WebdriverIO). Centers
   * the element in the viewport by default, useful when a sticky header/footer would otherwise
   * cover it.
   */
  public void jsScrollIntoView(WebElement element) {
    jsScrollIntoView(element, "center");
  }

  public void jsScrollIntoView(WebElement element, String block) {
    ((JavascriptExecutor) driver)
        .executeScript("arguments[0].scrollIntoView({block: arguments[1]});", element, block);
  }

  /**
   * WebElement#getCssValue already returns the browser-resolved value, so this is a thin wrapper
   * kept only for call-site consistency with the other site helpers.
   */
  public String getComputedStyle(WebElement element, String property) {
    return element.getCssValue(property);
  }

  /**
   * Strips {@code target="_blank"} before clicking so a link opens in the same tab/window,
   * avoiding window-handle juggling for sites that use target="_blank" pervasively.
   */
  public void openSameTab(WebElement link) {
    ((JavascriptExecutor) driver).executeScript("arguments[0].removeAttribute('target');", link);
    link.click();
  }
}
